add: initial support for Milk-V Jupiter/SpacemiT K1

Owen-sz opened the pull request:
Adds genimage.cfg, brings in vendor opensbi and u-boot via nix files. Adds Jupiter dev-image

Owen-sz committed:
add initial support for jupiter

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
add license headers to jupiter nix files

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
add license headers to jupiter cfg file

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
add juiter u-boot.txt to REUSE.toml

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
oops need comma

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
add license header to opensbi jupiter patch

Signed-off-by: Owen Zimmerman <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
Clean up unused lines in genimage file

Signed-off-by: Owen Zimmerman <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
fix jupiter package call line in nix flake

Signed-off-by: Owen <owen@fyralabs.com>
Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz commented:
Should be good now.

Owen-sz committed:
add jupiter to all-images

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
fix builds lol, specify genimage thingy

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
fix broken link outputs by renaming img output thing

Signed-off-by: Owen-sz <owen@fyralabs.com>

remexre commented on flake.nix:
```diff
@@ -10,12 +10,20 @@
       nixpkgs.follows = "nixpkgs";
     };
   };
+  inputs.u-boot-spacemit-k1 = {
```
Try instead doing e.g.

```
inputs = {
  u-boot-milkv-duos-sd = ...;
  u-boot-spacemit-k1 = ...;
};
```

remexre commented on src/image-milkv-jupiter/genimage.cfg:
```diff
@@ -0,0 +1,37 @@
+		size = 256K
+	}
+	partition env {
+		# When we can use genimage version 19, replace with 'u-boot-env,' version 19 will come with NixOS 22.11
```
File an issue to update this. Also update this.

remexre commented on src/image-milkv-jupiter/opensbi.nix:
```diff
@@ -0,0 +1,70 @@
+      # CHIP_ARCH=CV181X \
+      # FW_FDT_PATH=
+      # OPENSBI_PATH=$PWD \
+      # PLATFORM=generic
```
```suggestion
```

Owen-sz committed:
update nixos version needed for genimage

Signed-off-by: Owen-sz <owen@fyralabs.com>

remexre commented on src/image-milkv-jupiter/u-boot.txt:
```diff
@@ -0,0 +1,103 @@
+arch=riscv
```
Minimize this file. Probably you can just use the Milk-V Duo S one, changing `loadaddr` and `startaddr` to the values in this file.

remexre reviewed (commented):
Test that this boots the ukoOS kernel on real hardware, plz

Owen-sz committed:
remove comments in opensbi.nix

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
update u-boot.txt

Signed-off-by: Owen-sz <owen@fyralabs.com>

remexre commented on src/image-milkv-jupiter/genimage.cfg:
```diff
@@ -0,0 +1,37 @@
+		size = 256K
+	}
+	partition env {
+		# When we can use genimage version 19, replace with 'u-boot-env,' version 19 will come with NixOS 25.11
```
NixOS 25.11 came out lmao.

remexre commented on src/image-milkv-jupiter/u-boot.txt:
```diff
@@ -0,0 +1,20 @@
+autoload=0
```
needs testing on real hw; lmk when you did

Owen-sz replied:
It worky now

Owen-sz committed:
update partition type in jupiter genimage to u-boot-env

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
update nix flake

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz commented:
Notes for testing:

Setup:
Attach UART pins to CPU UART, not security chip UART thing

build:
`nix build -L .#all-images`

flash `src/result/dev-image-milkv-jupiter.img` to microSD card

serial into the board

Need to see what u-boot is doing (it's loading GRUB?)

pins to reset/ignore SPI flash, since u-boot is being loaded form SPI flash?

Owen-sz commented:
--- UPDATED NOTES ---

BOOT STATUS:

```
[   6.977] Trying to boot from SD...
[   7.090] switch to partitions #0, OK
[   7.090] mmc0 is current device
[   7.126] 136096 bytes read in 21 ms (6.2 MiB/s)
[   7.128] Unhandled exception: Store/AMO access fault
[   7.133] EPC: 000000007eea8ca8 RA: 000000007ef50ce0 TVAL: 0000000000000000
[   7.139] EPC: 0000000000201ca8 RA: 00000000002a9ce0 reloc adjusted
[   7.146] 
Code: b383 0385 be03 0405 be83 0485 bf03 0505 (e110)
[   7.151] 
```

CHANGES:

The commit after this will be updating the memory addresses in code to match that of the Jupiter, which is a temporary fix, and this should not be merged with those changes.
It also fixes the u-boot.txt file with more configs for the Jupiter.

TESTING:

Go to the ukoOS u-boot repo, use commit `ed29515c8baed05cdc1284474a847df214b0383c` from the spacemit-k1-testing branch, copy the `src/image-milkv-jupiter/u-boot.txt` from this Pull Request into the `board/spacemit/k1-x/k1-x.env`, then to build:
`nix build --override-input u-boot-spacemit-k1 ../ukoos-uboot/ -L .#all-images`

Owen-sz committed:
Align to Jupiter hardware

Correct Jupiter nix file to load kernel into correct spot
Add /boot into Jupiter genimage.cfg
Update u-boot.txt with Jupiter things, and set new startaddr because the Jupiter's RAM is not the same place as the Duo S and QEMU
Update files that rg told me had the Duo S/QEMU staraddr to temporarily to include the Jupiters startaddr

Signed-off-by: Owen-sz <owen@fyralabs.com>

jastintime commented on LICENSES/BSD-2-Clause.txt:
```diff
@@ -0,0 +1,22 @@
+Redistribution and use in source and binary forms, with or without
```
bruh why we bringing this bloat in

jastintime commented on src/image-milkv-jupiter/default.nix:
```diff
@@ -0,0 +1,62 @@
+# SPDX-FileCopyrightText: 2025 ukoOS Contributors
```
you can delete this, bloat

jastintime commented on src/image-milkv-jupiter/genimage.cfg:
```diff
@@ -0,0 +1,51 @@
+image boot.vfat {
+	vfat {
+		label = "ukoos-boot"
+	}
```
bloat wasting sizeof("ukoos-boot")

remexre replied:
ratio

jastintime commented on src/image-milkv-jupiter/opensbi.nix:
```diff
@@ -0,0 +1,65 @@
+  version = "2.2.27";
+
+  src = fetchgit {
+    url = "https://gitee.com/bianbu-linux/opensbi.git";
```
tell us what commit you are bringing in :+1:

jastintime replied:
keep up the good work!

Owen-sz replied:
Resolved in https://github.com/UMN-Kernel-Object/ukoos/pull/56/changes/bd0a87f091295356892abed404ed050ca80801e9

jastintime commented on src/image-milkv-jupiter/opensbi.nix:
```diff
@@ -0,0 +1,65 @@
+    make \
+      ''${enableParallelBuilding:+-j''${NIX_BUILD_CORES}} \
+      ARCH=riscv \
+      CROSS_COMPILE=riscv64-unknown-linux-musl- \
```
why linux-musl and not ukoos-nolibcs 😎

Owen-sz replied:
Resolved in https://github.com/UMN-Kernel-Object/ukoos/pull/56/commits/8ed801a524cf70f6f26b9476814d3ae60bd23408

Owen-sz replied:
Well, turns out using the embedded cross compiler breaks u-boot.

Hangs here:

```
U-Boot SPL 2022.10spacemit (Jan 01 1980 - 00:00:00 +0000)
[   0.247] DDR type LPDDR4X
[   0.247] set ddr tx odt to 80ohm!
[   0.262] lpddr4_silicon_init consume 15ms
[   0.263] Change DDR data rate to 2400MT/s
[   0.491] Boot from fit configuration k1-x_milkv-jupiter
[   0.493] ## Checking hash(es) for config conf_5 ... OK
[   0.498] ## Checking hash(es) for Image uboot ... crc32+ OK
[   0.509] ## Checking hash(es) for Image fdt_5 ... crc32+ OK
[   0.552] ## Checking hash(es) for config config_1 ... OK
[   0.555] ## Checking hash(es) for Image opensbi ... crc32+ OK
[   0.591] 
```

jastintime commented on src/image-milkv-jupiter/u-boot.txt:
```diff
@@ -0,0 +1,20 @@
+autoload=0
+autostart=no
+bootdelay=0
+bootfile=kernel.elf
+loadaddr=0x200000
+startaddr=0x8200000
```
appeared out of thin air 😢

Owen-sz replied:
We have found the proper startaddr. The upstream config defines `loadaddr=0x200000` (https://github.com/UMN-Kernel-Object/ukoos/pull/56/changes/bdd45d601f3c8069d1be7aae2cc48a86f8674977#diff-72e1a66e87454d3c767bfc90d5bc33c3d6f3229adbeddc970f4b56ada355c2eeR56).
If we subtract the `loadaddr` from `startaddr`, we get `0x8000000`, and also get exactly 128MiB of free space between `loadaddr` and `startaddr` for the `kernel.elf` file.

jastintime commented on src/image-milkv-jupiter/opensbi.nix:
```diff
@@ -0,0 +1,65 @@
+  nativeBuildInputs = [
+    dtc
+    pkgsCross.riscv64-musl.stdenv.cc.bintools.bintools
+    pkgsCross.riscv64-musl.stdenv.cc.cc
```
double check? why musl stuff

Owen-sz replied:
Resolved in https://github.com/UMN-Kernel-Object/ukoos/pull/56/commits/8ed801a524cf70f6f26b9476814d3ae60bd23408

Owen-sz replied:
Nevermind... https://github.com/UMN-Kernel-Object/ukoos/pull/56#discussion_r3142612732

jastintime commented on src/image-milkv-jupiter/genimage.cfg:
```diff
@@ -0,0 +1,51 @@
+		partition-type-uuid = "u-boot-env"
+		offset = "384K"
+		image = "uboot.env"
+		size = 256K
```
🫃 big size, checky

Owen-sz replied:
This matches the Ubuntu sizes

jastintime reviewed (changes requested):
look at my comments.

Appreciate your hard work and dedication, love to see it 🦾

Owen-sz committed:
Update flake.lock with latest hashes

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz commented:
It works!

When you get into u-boot on the Jupiter:

```
run custom_bootcmd
gok ${startaddr} ${fdtcontroladdr}
```

TODO: Make this automatic

Owen-sz committed:
Make bootstub startaddress variable in needed files so the correct startaddress is used per make target

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
run nix fmt

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz commented:
Resolves https://github.com/UMN-Kernel-Object/ukoos/issues/78

Owen-sz commented:
TODO: 

- gitignore bootstub.ld - DONE
- Go through more comments - DONE

remexre commented on src/kernel/arch/riscv64/generate_bootstub.py:
```diff
@@ -244,8 +244,10 @@ def main():
+    arg_parser.add_argument("bootstub_startaddr", type=lambda x: int(x,0))
     args = arg_parser.parse_args()
-
+    print(args.bootstub_startaddr)    
```
zap

Owen-sz committed:
uncommit bootstub.ld, add to .gitignore

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
remove printing the bootstub startaddr

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
pull in specific opensbi commit

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
use riscv64-none-elf-

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
Revert "use riscv64-none-elf-"

This reverts commit 8ed801a524cf70f6f26b9476814d3ae60bd23408.

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
comment about why we are using musl compilers

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
properly use our own u-boot config

remove mkenvimage as we are using env import -t in the u-boot repo now

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
remove env partition, update flake.lock, indent in genimage.cfg properly

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz commented:
Update - To run kernel once in the Jupiter, run `run bootcmd`

Owen-sz committed:
update flake.lock

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
make the opensbi patch actually apply

Signed-off-by: Owen-sz <owen@fyralabs.com>

remexre commented on src/kernel/arch/riscv64/generate_bootstub.py:
```diff
@@ -244,8 +244,9 @@ def main():
+    arg_parser.add_argument("bootstub_startaddr", type=lambda x: int(x,0))
     args = arg_parser.parse_args()
-
+    
```
```suggestion

```

Owen-sz replied:
Resolved in https://github.com/UMN-Kernel-Object/ukoos/pull/56/changes/d977e337c8a3aa0d416448817505cf0d5d738e30

Owen-sz committed:
fix tab that was added

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
add u-boot building link

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
fix genimage formatting

Signed-off-by: Owen-sz <owen@fyralabs.com>

Owen-sz committed:
update comment about u-boot compile issue

Signed-off-by: Owen-sz <owen@fyralabs.com>

jastintime reviewed (approved):
lgtm, great work 👍
